File: src/lib/data/mappers.ts
import { DEFAULT_NOTIFICATION_PREFERENCES } from '@/lib/defaults'
import type {
  Account,
  AppNotification,
  Budget,
  Category,
  Goal,
  NotificationPreferences,
  Profile,
  RecurringTransaction,
  SavedFilter,
  Transaction,
} from '@/types'

export function mapProfile(row: Record<string, unknown>, email: string): Profile {
  return {
    id: String(row.id),
    email,
    fullName: String(row.full_name ?? ''),
    avatarUrl: (row.avatar_url as string | null) ?? null,
    currency: String(row.currency ?? 'USD'),
    dateFormat: String(row.date_format ?? 'MMM d, yyyy'),
    theme: (row.theme as Profile['theme']) ?? 'system',
    onboardingCompleted: Boolean(row.onboarding_completed),
    notificationPreferences: {
      ...DEFAULT_NOTIFICATION_PREFERENCES,
      ...((row.notification_preferences as NotificationPreferences | null) ?? {}),
    },
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapAccount(row: Record<string, unknown>): Account {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    name: String(row.name),
    type: row.type as Account['type'],
    openingBalance: Number(row.opening_balance),
    currency: String(row.currency),
    icon: String(row.icon),
    color: String(row.color),
    status: row.status as Account['status'],
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapCategory(row: Record<string, unknown>): Category {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    name: String(row.name),
    kind: row.kind as Category['kind'],
    icon: String(row.icon),
    color: String(row.color),
    parentId: (row.parent_id as string | null) ?? null,
    sortOrder: Number(row.sort_order ?? 0),
    isSystem: Boolean(row.is_system),
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapTransaction(row: Record<string, unknown>): Transaction {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    type: row.type as Transaction['type'],
    amount: Number(row.amount),
    currency: String(row.currency),
    categoryId: (row.category_id as string | null) ?? null,
    subcategoryId: (row.subcategory_id as string | null) ?? null,
    accountId: String(row.account_id),
    toAccountId: (row.to_account_id as string | null) ?? null,
    merchant: String(row.merchant ?? ''),
    description: String(row.description ?? ''),
    notes: String(row.notes ?? ''),
    date: String(row.date),
    paymentMethod: row.payment_method as Transaction['paymentMethod'],
    tags: (row.tags as string[]) ?? [],
    recurringId: (row.recurring_id as string | null) ?? null,
    attachmentPath: (row.attachment_path as string | null) ?? null,
    attachmentName: (row.attachment_name as string | null) ?? null,
    isSample: Boolean(row.is_sample),
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapBudget(row: Record<string, unknown>): Budget {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    name: String(row.name),
    categoryId: (row.category_id as string | null) ?? null,
    limitAmount: Number(row.limit_amount),
    period: row.period as Budget['period'],
    startDate: String(row.start_date),
    endDate: (row.end_date as string | null) ?? null,
    alertThreshold: Number(row.alert_threshold),
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapRecurring(row: Record<string, unknown>): RecurringTransaction {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    type: row.type as RecurringTransaction['type'],
    amount: Number(row.amount),
    currency: String(row.currency),
    categoryId: (row.category_id as string | null) ?? null,
    accountId: String(row.account_id),
    merchant: String(row.merchant ?? ''),
    notes: String(row.notes ?? ''),
    paymentMethod: row.payment_method as RecurringTransaction['paymentMethod'],
    frequency: row.frequency as RecurringTransaction['frequency'],
    interval: Number(row.interval ?? 1),
    startDate: String(row.start_date),
    endDate: (row.end_date as string | null) ?? null,
    nextOccurrence: String(row.next_occurrence),
    active: Boolean(row.active),
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapGoal(row: Record<string, unknown>): Goal {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    name: String(row.name),
    targetAmount: Number(row.target_amount),
    currentAmount: Number(row.current_amount),
    deadline: (row.deadline as string | null) ?? null,
    icon: String(row.icon),
    color: String(row.color),
    createdAt: String(row.created_at),
    updatedAt: String(row.updated_at),
  }
}

export function mapNotification(row: Record<string, unknown>): AppNotification {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    type: row.type as AppNotification['type'],
    title: String(row.title),
    body: String(row.body),
    read: Boolean(row.read),
    metadata: (row.metadata as Record<string, string>) ?? {},
    createdAt: String(row.created_at),
  }
}

export function mapSavedFilter(row: Record<string, unknown>): SavedFilter {
  return {
    id: String(row.id),
    userId: String(row.user_id),
    name: String(row.name),
    filters: (row.filters as SavedFilter['filters']) ?? {},
    createdAt: String(row.created_at),
  }
}

export function transactionToRow(input: Partial<Transaction>, userId?: string) {
  return {
    ...(userId ? { user_id: userId } : {}),
    type: input.type,
    amount: input.amount,
    currency: input.currency,
    category_id: input.categoryId ?? null,
    subcategory_id: input.subcategoryId ?? null,
    account_id: input.accountId,
    to_account_id: input.toAccountId ?? null,
    merchant: input.merchant ?? '',
    description: input.description ?? '',
    notes: input.notes ?? '',
    date: input.date,
    payment_method: input.paymentMethod,
    tags: input.tags ?? [],
    recurring_id: input.recurringId ?? null,
    attachment_path: input.attachmentPath ?? null,
    attachment_name: input.attachmentName ?? null,
    is_sample: input.isSample ?? false,
  }
}

File: flutter/lib/data/mappers.dart
import '../core/defaults.dart';
import '../models/models.dart';

String? _string(Object? value) => value?.toString();

double _number(Object? value) => value is num ? value.toDouble() : double.tryParse('$value') ?? 0;

int _int(Object? value) => value is num ? value.toInt() : int.tryParse('$value') ?? 0;

bool _bool(Object? value) => value == true;

Map<String, dynamic> _map(Object? value) =>
    value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};

const camelToSnake = {
  'userId': 'user_id',
  'fullName': 'full_name',
  'avatarUrl': 'avatar_url',
  'dateFormat': 'date_format',
  'onboardingCompleted': 'onboarding_completed',
  'notificationPreferences': 'notification_preferences',
  'openingBalance': 'opening_balance',
  'parentId': 'parent_id',
  'sortOrder': 'sort_order',
  'isSystem': 'is_system',
  'categoryId': 'category_id',
  'subcategoryId': 'subcategory_id',
  'accountId': 'account_id',
  'toAccountId': 'to_account_id',
  'paymentMethod': 'payment_method',
  'recurringId': 'recurring_id',
  'attachmentPath': 'attachment_path',
  'attachmentName': 'attachment_name',
  'isSample': 'is_sample',
  'limitAmount': 'limit_amount',
  'startDate': 'start_date',
  'endDate': 'end_date',
  'alertThreshold': 'alert_threshold',
  'nextOccurrence': 'next_occurrence',
  'targetAmount': 'target_amount',
  'currentAmount': 'current_amount',
  'createdAt': 'created_at',
  'updatedAt': 'updated_at',
};

const _identityKeys = {'id', 'userId', 'createdAt', 'updatedAt', 'email'};

Map<String, dynamic> camelPatchToRow(
  Map<String, dynamic> patch, {
  String? userId,
  bool dropIdentity = true,
}) {
  final row = <String, dynamic>{};
  if (userId != null) row['user_id'] = userId;
  patch.forEach((key, value) {
    if (dropIdentity && _identityKeys.contains(key)) return;
    row[camelToSnake[key] ?? key] = value;
  });
  return row;
}

NotificationPreferences mapNotificationPreferences(Object? raw) {
  final map = _map(raw);
  return NotificationPreferences(
    budgetAlerts: map['budgetAlerts'] as bool? ?? map['budget_alerts'] as bool? ?? defaultNotificationPreferences.budgetAlerts,
    recurringAlerts:
        map['recurringAlerts'] as bool? ?? map['recurring_alerts'] as bool? ?? defaultNotificationPreferences.recurringAlerts,
    goalAlerts: map['goalAlerts'] as bool? ?? map['goal_alerts'] as bool? ?? defaultNotificationPreferences.goalAlerts,
    importExportAlerts: map['importExportAlerts'] as bool? ??
        map['import_export_alerts'] as bool? ??
        defaultNotificationPreferences.importExportAlerts,
  );
}

Profile mapProfile(Map<String, dynamic> row, String email) {
  return Profile(
    id: '${row['id']}',
    email: email,
    fullName: _string(row['full_name']) ?? '',
    avatarUrl: _string(row['avatar_url']),
    currency: _string(row['currency']) ?? 'USD',
    dateFormat: _string(row['date_format']) ?? 'MMM d, yyyy',
    theme: ThemePreferenceX.parse(_string(row['theme']) ?? 'system'),
    onboardingCompleted: _bool(row['onboarding_completed']),
    notificationPreferences: mapNotificationPreferences(row['notification_preferences']),
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

Account mapAccount(Map<String, dynamic> row) {
  return Account(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    name: _string(row['name']) ?? '',
    type: AccountTypeX.parse(_string(row['type']) ?? 'bank'),
    openingBalance: _number(row['opening_balance']),
    currency: _string(row['currency']) ?? 'USD',
    icon: _string(row['icon']) ?? 'Landmark',
    color: _string(row['color']) ?? '#3B82F6',
    status: AccountStatusX.parse(_string(row['status']) ?? 'active'),
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

Category mapCategory(Map<String, dynamic> row) {
  return Category(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    name: _string(row['name']) ?? '',
    kind: CategoryKindX.parse(_string(row['kind']) ?? 'expense'),
    icon: _string(row['icon']) ?? 'CircleEllipsis',
    color: _string(row['color']) ?? '#94A3B8',
    parentId: _string(row['parent_id']),
    sortOrder: _int(row['sort_order']),
    isSystem: _bool(row['is_system']),
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

Transaction mapTransaction(Map<String, dynamic> row) {
  return Transaction(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    type: TransactionTypeX.parse(_string(row['type']) ?? 'expense'),
    amount: _number(row['amount']),
    currency: _string(row['currency']) ?? 'USD',
    categoryId: _string(row['category_id']),
    subcategoryId: _string(row['subcategory_id']),
    accountId: '${row['account_id']}',
    toAccountId: _string(row['to_account_id']),
    merchant: _string(row['merchant']) ?? '',
    description: _string(row['description']) ?? '',
    notes: _string(row['notes']) ?? '',
    date: _string(row['date']) ?? '',
    paymentMethod: PaymentMethodX.parse(_string(row['payment_method']) ?? 'card'),
    tags: (row['tags'] as List?)?.map((item) => '$item').toList() ?? const [],
    recurringId: _string(row['recurring_id']),
    attachmentPath: _string(row['attachment_path']),
    attachmentName: _string(row['attachment_name']),
    isSample: _bool(row['is_sample']),
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

Budget mapBudget(Map<String, dynamic> row) {
  return Budget(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    name: _string(row['name']) ?? '',
    categoryId: _string(row['category_id']),
    limitAmount: _number(row['limit_amount']),
    period: BudgetPeriodX.parse(_string(row['period']) ?? 'monthly'),
    startDate: _string(row['start_date']) ?? '',
    endDate: _string(row['end_date']),
    alertThreshold: _number(row['alert_threshold']),
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

RecurringTransaction mapRecurring(Map<String, dynamic> row) {
  return RecurringTransaction(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    type: TransactionTypeX.parse(_string(row['type']) ?? 'expense'),
    amount: _number(row['amount']),
    currency: _string(row['currency']) ?? 'USD',
    categoryId: _string(row['category_id']),
    accountId: '${row['account_id']}',
    merchant: _string(row['merchant']) ?? '',
    notes: _string(row['notes']) ?? '',
    paymentMethod: PaymentMethodX.parse(_string(row['payment_method']) ?? 'card'),
    frequency: RecurrenceFrequencyX.parse(_string(row['frequency']) ?? 'monthly'),
    interval: _int(row['interval'] ?? 1),
    startDate: _string(row['start_date']) ?? '',
    endDate: _string(row['end_date']),
    nextOccurrence: _string(row['next_occurrence']) ?? '',
    active: row['active'] == false ? false : true,
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

Goal mapGoal(Map<String, dynamic> row) {
  return Goal(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    name: _string(row['name']) ?? '',
    targetAmount: _number(row['target_amount']),
    currentAmount: _number(row['current_amount']),
    deadline: _string(row['deadline']),
    icon: _string(row['icon']) ?? 'Target',
    color: _string(row['color']) ?? '#A855F7',
    createdAt: _string(row['created_at']) ?? '',
    updatedAt: _string(row['updated_at']) ?? '',
  );
}

AppNotification mapNotification(Map<String, dynamic> row) {
  return AppNotification(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    type: NotificationTypeX.parse(_string(row['type']) ?? 'budget_exceeded'),
    title: _string(row['title']) ?? '',
    body: _string(row['body']) ?? '',
    read: _bool(row['read']),
    metadata: _map(row['metadata']).map((key, value) => MapEntry(key, '$value')),
    createdAt: _string(row['created_at']) ?? '',
  );
}

SavedFilter mapSavedFilter(Map<String, dynamic> row) {
  return SavedFilter(
    id: '${row['id']}',
    userId: '${row['user_id']}',
    name: _string(row['name']) ?? '',
    filters: TransactionFilters.fromJson(_map(row['filters'])),
    createdAt: _string(row['created_at']) ?? '',
  );
}

Map<String, dynamic> transactionToRow(Map<String, dynamic> input, [String? userId]) {
  return camelPatchToRow(input, userId: userId);
}
